### PR TITLE
Daemon shutdown restart

### LINKED ISSUE #30: [User Story] As a systems engineer, I want the Daemon to shut down and restart if it uses more than 10% CPU or 1Mb memory
### **Is your user story related to a problem? Please describe.**
It frustrates me when a background process uses too much CPU or memory and does not fix itself.

### **Describe the solution you'd like**
The daemon should restart should either of these conditions be met.

### **Describe alternatives you've considered**
Restarting is the most effective way to fix CPU and memory usage issues.

### **Additional context**
The daemon does not need both CPU and Memory usage issues to restart. I should restart when it encounters one or the other.

### PR DESCRIPTION
**Pull request by arich94**
This checks that the daemon process is using less than or equal to 10% CPU and 10MB of physical memory. If it is more than either of these, the daemon will shut down. Restart process already integrated previously.

Closes #30  #29 #28

### CODE REVIEW

File: daemon/src/system_report.py [hunk 1]
```diff
@@ -89,9 +89,14 @@ def print_report(self):
     def add_system_process_info(self):
         process_list = list()
         process_buffer = list()
+        name_pattern = "python3? system_report.py"
```
Comment (Pyxsys) [inline daemon/src/system_report.py:92]: regex pattern is incorrect to expected name. 
Current regex would capture "python3? system_reporthpy" and not the possible "python3 /app/bin/system_report.py /config/file/path"

An example of an acceptable regex would be `python(?:\d|)\W+(?:.*)system_report\.py`

File: daemon/src/system_report.py [hunk 2]
```diff
@@ -89,9 +89,14 @@ def print_report(self):
...
 
         #Initialize start time for process diagnostic scan
         for proc in SysScrubber.fetch_all_processes():
+            process_name = SysScrubber.fetch_process_name(proc.pid)
+            if re.search(name_pattern, process_name):
+                if proc.cpu_percent() > 10 or proc.memory_info()[0] > 1000000:
```
Comment (Pyxsys) [inline daemon/src/system_report.py:98]: bind values to a reference variable. Ideally sourced from configuration file.

**Review by Pyxsys — CHANGES_REQUESTED** [review]
The below needs to be addressed:

- [ ] Revise regex to capture correct process name (verify in virtual linux environment)
- [ ] Remove hard-coded boundaries and bind them to configurable variables.
- [ ] Revise solution to cover more of the flow, not only in add_system_process_info()
- [ ] Add a meaningful indicator that the flow terminated due to exceeding permitted bounds. (e.g. print something to log)
- [ ] Add tests

For the third point we can consider spawning an "Observer-Thread" that periodically checks the resource consumption. A custom class can be created that extends the python Threading class to accomplish this.

File: daemon/src/system_report.py (hunk 1 above)
Comment (arich94) [reply daemon/src/system_report.py]: I used a similar form to the one you suggested, but is a little bit simpler.

File: daemon/src/system_report.py (hunk 2 above)
Comment (arich94) [reply daemon/src/system_report.py]: I added the variables to the json configuration file. The values can now be edited there.

**Comment by arich94** [comment]
> The below needs to be addressed:
> 
> * [x]  Revise regex to capture correct process name (verify in virtual linux environment)
> * [x]  Remove hard-coded boundaries and bind them to configurable variables.
> * [x]  Revise solution to cover more of the flow, not only in add_system_process_info()
> * [x]  Add a meaningful indicator that the flow terminated due to exceeding permitted bounds. (e.g. print something to log)
> * [x]  Add tests
> 
> For the third point we can consider spawning an "Observer-Thread" that periodically checks the resource consumption. A custom class can be created that extends the python Threading class to accomplish this.

Unfortuantely, the daemon program cannot be closed from a thread. This issue was resolved by having the daemon's scrubbing functions be a thread, and the checks for cpu and memory usage be made from the main thread, where the program can be closed. This makes it so that the checks can be consistent throughout the execution of the runner.

File: daemon/src/system_report.py [hunk 3]
```diff
@@ -603,16 +621,50 @@ def fetch_net_adapter_addrs(cls, adapter_name = None):
...
+class DaemonChecker:
+    def __init__(self, path):
+        with open(path, "r") as config_file:
+            self.configs = json.load(config_file)
+
+    def check_daemon(self):
+        name_pattern = "python(\d?)\W+(.*)system_report\.py"
+        
+        if self.configs["max_cpu"] <= 1 or self.configs["max_memory"] <= 1:
+            return
```
Comment (Pyxsys) [inline daemon/src/system_report.py:633]: What is the purpose of this if-statement?
It reads as if the configurations for either metric is a floating value less than or equal to one (eg. 0, 1, 0.3141567, -140882) to then skip the rest of the checks. 
Is the range of values intended?

**Review by Pyxsys — CHANGES_REQUESTED** [review]
Requested changes were made.
Threading solution is acceptable.

Please address if-statement check concern.

File: daemon/src/system_report.py (hunk 3 above)
Comment (arich94) [reply daemon/src/system_report.py]: Bounds have been set to 0 as they should be. Marking this conversation as resolved.

Comment (Pyxsys) [reply daemon/src/system_report.py]: In-line with your reply, please update the bounds to ignore to be if less than or equal to 0 instead of 1.

**Review by DerekRuiz — APPROVED** [review]
New Classes looks good
All code seems tidy

**Pull request merged**